video/out/drm_prime.c

```diff
@@ -78,15 +81,71 @@ int drm_prime_create_framebuffer(struct mp_log *log, int fd, AVDRMFrameDescripto
    return -1;
 }
 
-void drm_prime_destroy_framebuffer(struct mp_log *log, int fd, struct  drm_prime_framebuffer *framebuffer)
+void drm_prime_destroy_framebuffer(struct mp_log *log, int fd,
+                                    struct drm_prime_framebuffer *framebuffer,
+                                    struct drm_prime_handle_refs *handle_refs)
 {
     if (framebuffer->fb_id)
         drmModeRmFB(fd, framebuffer->fb_id);
 
     for (int i = 0; i < AV_DRM_MAX_PLANES; i++) {
-        if (framebuffer->gem_handles[i])
-            drmIoctl(fd, DRM_IOCTL_GEM_CLOSE, &framebuffer->gem_handles[i]);
+        if (framebuffer->gem_handles[i]) {
+            drm_prime_remove_handle_ref(handle_refs,
+                                         framebuffer->gem_handles[i]);
+            if(!drm_prime_get_handle_ref_count(handle_refs, framebuffer->gem_handles[i])) {
+                drmIoctl(fd, DRM_IOCTL_GEM_CLOSE, &framebuffer->gem_handles[i]);
+            }
+        }
     }
 
     memset(framebuffer, 0, sizeof(*framebuffer));
 }
+
+void drm_prime_init_handle_ref_count(struct drm_prime_handle_refs *handle_refs) {
+    //4 dumb buffer + (1 per plane * planes we keep) + 10 (extra safety)
+    uint16_t init_size = 4 + (AV_DRM_MAX_PLANES * 4) + 10;
+    handle_refs->handle_ref_count = (uint32_t *)malloc(init_size * sizeof(uint32_t));
+    handle_refs->size = init_size;
+
+    memset(handle_refs->handle_ref_count, 0, sizeof(uint32_t) * init_size);
+}
+
+void drm_prime_add_handle_ref(struct drm_prime_handle_refs *handle_refs, uint32_t handle) {
+    if(handle) {
```

Comment:
Isn't this a file handle? Valid file handles can be 0. That may be unlikely, since it will be either occupied by stdin, or anything else that was opened before, but still.

Reply:
I looked it up in the implemenation of the kernel. It is allocated by idr_alloc and starts at 1.

https://github.com/torvalds/linux/blob/776e49e8ddb5169e6477fd33a396e9c7b2eb7400/drivers/gpu/drm/drm_gem.c#L392

Reply:
Oh OK, so not a file handle at all. Is still guaranteed to reuse IDs in the lower integer range? Since you're directly mapping them to array indexes.

Reply:
If I understand you correctly, yes idr looks for the next free number starting from 1. So if 2 gets freed the next requesting a handle number gets the 2.